subsys/net/ip/net_if.c

```diff
@@ -2151,6 +2151,47 @@ struct net_if_mcast_addr *net_if_ipv4_maddr_lookup(const struct in_addr *maddr,
 }
 #endif /* CONFIG_NET_IPV4 */
 
+enum net_verdict net_if_recv_data(struct net_if *iface, struct net_pkt *pkt)
+{
+	if (IS_ENABLED(CONFIG_NET_PROMISCUOUS_MODE) &&
+	    net_if_is_promisc(iface)) {
+		/* If the packet is not for us and the promiscuous
+		 * mode is enabled, then increase the ref count so
+		 * that net_core.c:processing_data() will not free it.
+		 * The promiscuous mode handler must free the packet
+		 * after it has finished working with it.
+		 *
+		 * If packet is for us, then NET_CONTINUE is returned.
+		 * In this case we must clone the packet, as the packet
```

Comment:
Why all these complications? Why not call promisc handler first, and make a stipulation that it should not alter packet (if it needs to, it should clone it and operate on clone).

Reply:
There is a pipe between core stack and application for these promiscuous packets which means that packet might be destroyed already when application gets it. So in order to avoid any issue, it is cloned here.